### PR TITLE
default_policy: Fix deadlock on LatencyAwareness used

### PR DESCRIPTION
**Pull request by wprzytula**
Described in #696.
Rust, unfortunately, makes lock guards taken on right side of `if let` expression live for else branch duration, too. This results in taking write lock while holding read lock.

In order to compile, I had to hint the `usize` type in one unrelated place, hence that change.

Added a regression test for the issue.

Fixes: #696

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/transport/load_balancing/default.rs [hunk 1]
```diff
@@ -1619,16 +1619,27 @@ mod latency_awareness {
         }
 
         pub(super) fn report_query(&self, node: &Node, latency: Duration) {
-            if let Some(node_avg) = self.node_avgs.read().unwrap().get(&node.host_id) {
+            let guard = self.node_avgs.read().unwrap();
```
Comment (havaker) [inline scylla/src/transport/load_balancing/default.rs:1622]: How about naming it `node_avgs_guard`?

**Review by havaker — COMMENTED** [review]
LGTM

File: scylla/src/transport/load_balancing/default.rs [hunk 2]
```diff
@@ -757,7 +757,7 @@ mod tests {
                 for got in gots {
                     // First, make sure that `got` has the right number of items,
                     // equal to the sum of sizes of all expected groups
-                    let combined_groups_len = self.groups.iter().map(|s| s.len()).sum();
+                    let combined_groups_len: usize = self.groups.iter().map(|s| s.len()).sum();
```
Comment (cvybhu) [inline scylla/src/transport/load_balancing/default.rs:760]: I removed the `usize` and it still compiles with rust 1.68, but I'm a fan of having explicit type annotations so we could leave it as is.

**Review by cvybhu — APPROVED** [review]

**Review by cvybhu — APPROVED** [review]

File: scylla/src/transport/load_balancing/default.rs [hunk 3]
```diff
@@ -2582,5 +2594,29 @@ mod latency_awareness {
...
+                .into_handle();
+
+            let session = SessionBuilder::new()
+                .known_node(uri)
+                .default_execution_profile_handle(handle)
+                .build()
+                .await
+                .unwrap();
+
+            session.query("whatever", ()).await.unwrap_err();
```
Comment (piodul) [inline scylla/src/transport/load_balancing/default.rs:2619]: This statement should send a valid query and check that it succeeded. If it starts failing due to some unrelated reason the test will stop being useful.

Comment (wprzytula) [reply scylla/src/transport/load_balancing/default.rs:2619]: This test was meant as a regression test, so it only tests that the test completes. We could write some other sophisticated integration tests to ensure that LAP does not cause unpredictable behaviour.

**Pull request merged**